File: Coding/python/hasher.py
"""Idiomatic Python 3 port of java/Hasher.java (original package code.ds).

Three data structures:
  * Hasher        - generic hash table with separate chaining (doubly-linked
                    nodes per bucket)
  * MiniCassandra - two-level key/value store with ordered column ranges
  * ConsistentHash - consistent-hashing ring (with the source's stub hash that
                     always returns 0)

Java generics ``<K, V>`` become plain Python. ``TreeMap`` (a sorted map) is
emulated with a regular ``dict`` whose keys are sorted on demand. The
``HashFunction.hash`` method in the source always returns 0; that behaviour is
preserved, so every node maps to the same ring slot.
"""


# ---------------------------------------------------------------------------
# Hasher: separate-chaining hash table
# ---------------------------------------------------------------------------
class LinkedListNode:
    def __init__(self, k, v):
        self.next = None
        self.prev = None
        self.key = k
        self.value = v

    def print_forward(self):
        data = "({},{})".format(self.key, self.value)
        if self.next is not None:
            return data + "->" + self.next.print_forward()
        return data


class Hasher:
    def __init__(self, capacity):
        # List of bucket heads (each a linked list), all initially empty.
        self.arr = [None] * capacity

    def put(self, key, value):
        node = self._get_node_for_key(key)
        if node is not None:
            old_value = node.value
            node.value = value  # just update the value
            return old_value

        node = LinkedListNode(key, value)
        index = self.get_index_for_key(key)
        if self.arr[index] is not None:
            node.next = self.arr[index]
            node.next.prev = node
        self.arr[index] = node
        return None

    def remove(self, key):
        node = self._get_node_for_key(key)
        if node is None:
            return None

        if node.prev is not None:
            node.prev.next = node.next
        else:
            # Removing head - update bucket.
            hash_key = self.get_index_for_key(key)
            self.arr[hash_key] = node.next

        if node.next is not None:
            node.next.prev = node.prev
        return node.value

    def get(self, key):
        if key is None:
            return None
        node = self._get_node_for_key(key)
        return None if node is None else node.value

    def _get_node_for_key(self, key):
        index = self.get_index_for_key(key)
        current = self.arr[index]
        while current is not None:
            if current.key == key:
                return current
            current = current.next
        return None

    def get_index_for_key(self, key):
        # "Really stupid function to map a key to an index."
        return abs(hash(key) % len(self.arr))

    def print_table(self):
        for i in range(len(self.arr)):
            s = "" if self.arr[i] is None else self.arr[i].print_forward()
            print("{}: {}".format(i, s))


# ---------------------------------------------------------------------------
# MiniCassandra: two-level keyed store (raw_key -> column_key -> value)
# ---------------------------------------------------------------------------
class Column:
    def __init__(self, key, value):
        self.key = key
        self.value = value

    def __repr__(self):
        return "Column({},{})".format(self.key, self.value)


class MiniCassandra:
    def __init__(self):
        self.map = {}

    def insert(self, raw_key, column_key, column_value):
        tm = self.map.get(raw_key)
        if tm is None:
            tm = {}
            self.map[raw_key] = tm
        tm[column_key] = column_value

    def query(self, raw_key, column_start, column_end):
        results = []
        tm = self.map.get(raw_key)
        if tm is None:
            return results
        # TreeMap.subMap(start, true, end, true): inclusive on both ends,
        # returned in ascending key order.
        for key in sorted(tm.keys()):
            if column_start <= key <= column_end:
                results.append(Column(key, tm[key]))
        return results


# ---------------------------------------------------------------------------
# ConsistentHash: hashing ring
# ---------------------------------------------------------------------------
class HashFunction:
    def hash(self, key):
        return 0  # faithful stub from the source


class ConsistentHash:
    def __init__(self, hash_function, number_of_replicas, nodes):
        self.hash_function = hash_function
        self.number_of_replicas = number_of_replicas
        self.circle = {}  # emulated SortedMap<Integer, T>
        for node in nodes:
            self.add(node)

    def add(self, node):
        for i in range(self.number_of_replicas):
            self.circle[self.hash_function.hash(str(node) + str(i))] = node

    def remove(self, node):
        for i in range(self.number_of_replicas):
            self.circle.pop(self.hash_function.hash(str(node) + str(i)), None)

    def get(self, key):
        if not self.circle:
            return None
        h = self.hash_function.hash(key)
        if h not in self.circle:
            keys = sorted(self.circle.keys())
            tail = [k for k in keys if k >= h]  # circle.tailMap(hash)
            h = tail[0] if tail else keys[0]
        return self.circle[h]


if __name__ == "__main__":
    print("Hasher")
    print()

    h = Hasher(10)
    h.put(1, "one")
    h.put(11, "eleven")  # collides with key 1 (both index 1) -> chained
    h.put(2, "two")
    print("get(1):", h.get(1), " get(11):", h.get(11), " get(2):", h.get(2))
    print("remove(1):", h.remove(1), " get(1) after remove:", h.get(1))

    mc = MiniCassandra()
    mc.insert("user1", 3, "c")
    mc.insert("user1", 1, "a")
    mc.insert("user1", 2, "b")
    print("MiniCassandra query user1 [1,2]:", mc.query("user1", 1, 2))

    ch = ConsistentHash(HashFunction(), 3, ["nodeA", "nodeB"])
    print("ConsistentHash get('x'):", ch.get("x"))

File: Coding/golang/hasher/hasher.go
package main

// Ported from Coding/java/Hasher.java (original package code.ds).
// Hash table with separate chaining + MiniCassandra + ConsistentHash.

import (
	"fmt"
	"sort"
)

func main() {
	fmt.Print("Hasher")
}

func absInt(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

// llNode is a doubly-linked-list node used for separate chaining.
type llNode[K comparable, V any] struct {
	next  *llNode[K, V]
	prev  *llNode[K, V]
	key   K
	value V
}

func (n *llNode[K, V]) printForward() string {
	data := fmt.Sprintf("(%v,%v)", n.key, n.value)
	if n.next != nil {
		return data + "->" + n.next.printForward()
	}
	return data
}

// Hasher is a hash table with separate chaining. A hash function for K must be
// supplied (Go has no built-in Object.hashCode for arbitrary keys).
type Hasher[K comparable, V any] struct {
	arr    []*llNode[K, V]
	hashFn func(K) int
}

func NewHasher[K comparable, V any](capacity int, hashFn func(K) int) *Hasher[K, V] {
	h := &Hasher[K, V]{hashFn: hashFn}
	h.arr = make([]*llNode[K, V], capacity)
	return h
}

// put inserts key/value, returning the previous value (zero value if new).
func (h *Hasher[K, V]) put(key K, value V) V {
	var zero V
	node := h.getNodeForKey(key)
	if node != nil {
		oldValue := node.value
		node.value = value // just update the value.
		return oldValue
	}

	node = &llNode[K, V]{key: key, value: value}
	index := h.getIndexForKey(key)
	if h.arr[index] != nil {
		node.next = h.arr[index]
		node.next.prev = node
	}
	h.arr[index] = node
	return zero
}

// remove deletes the node for key and returns its value (zero value if absent).
func (h *Hasher[K, V]) remove(key K) V {
	var zero V
	node := h.getNodeForKey(key)
	if node == nil {
		return zero
	}

	if node.prev != nil {
		node.prev.next = node.next
	} else {
		// Removing head - update.
		hashKey := h.getIndexForKey(key)
		h.arr[hashKey] = node.next
	}

	if node.next != nil {
		node.next.prev = node.prev
	}
	return node.value
}

func (h *Hasher[K, V]) get(key K) V {
	var zero V
	node := h.getNodeForKey(key)
	if node == nil {
		return zero
	}
	return node.value
}

func (h *Hasher[K, V]) getNodeForKey(key K) *llNode[K, V] {
	index := h.getIndexForKey(key)
	current := h.arr[index]
	for current != nil {
		if current.key == key {
			return current
		}
		current = current.next
	}
	return nil
}

func (h *Hasher[K, V]) getIndexForKey(key K) int {
	return absInt(h.hashFn(key) % len(h.arr))
}

func (h *Hasher[K, V]) printTable() {
	for i := 0; i < len(h.arr); i++ {
		s := ""
		if h.arr[i] != nil {
			s = h.arr[i].printForward()
		}
		fmt.Printf("%d: %s\n", i, s)
	}
}

// MiniCassandra is a two-level key/value store (raw_key -> column_key -> value).
type Column struct {
	key   int
	value string
}

func newColumn(key int, value string) *Column {
	return &Column{key: key, value: value}
}

type MiniCassandra struct {
	m map[string]map[int]string
}

func newMiniCassandra() *MiniCassandra {
	return &MiniCassandra{m: make(map[string]map[int]string)}
}

func (c *MiniCassandra) insert(rawKey string, columnKey int, columnValue string) {
	tm := c.m[rawKey]
	if tm == nil {
		tm = make(map[int]string)
		c.m[rawKey] = tm
	}
	tm[columnKey] = columnValue
}

// query returns columns in the inclusive range [columnStart, columnEnd], sorted by key.
func (c *MiniCassandra) query(rawKey string, columnStart, columnEnd int) []*Column {
	results := []*Column{}
	tm := c.m[rawKey]
	if tm == nil {
		return results
	}
	keys := []int{}
	for k := range tm {
		if k >= columnStart && k <= columnEnd {
			keys = append(keys, k)
		}
	}
	sort.Ints(keys)
	for _, k := range keys {
		results = append(results, newColumn(k, tm[k]))
	}
	return results
}

// hashFunction is a placeholder hash (matches the Java stub which returns 0).
type hashFunction struct{}

func (hashFunction) hash(key interface{}) int {
	return 0
}

// ConsistentHash maps nodes around a hash ring with virtual replicas.
type ConsistentHash[T any] struct {
	hashFunction     hashFunction
	numberOfReplicas int
	circle           map[int]T
}

func newConsistentHash[T any](hf hashFunction, numberOfReplicas int, nodes []T) *ConsistentHash[T] {
	c := &ConsistentHash[T]{
		hashFunction:     hf,
		numberOfReplicas: numberOfReplicas,
		circle:           make(map[int]T),
	}
	for _, node := range nodes {
		c.add(node)
	}
	return c
}

func (c *ConsistentHash[T]) add(node T) {
	for i := 0; i < c.numberOfReplicas; i++ {
		c.circle[c.hashFunction.hash(fmt.Sprintf("%v%d", node, i))] = node
	}
}

func (c *ConsistentHash[T]) remove(node T) {
	for i := 0; i < c.numberOfReplicas; i++ {
		delete(c.circle, c.hashFunction.hash(fmt.Sprintf("%v%d", node, i)))
	}
}

func (c *ConsistentHash[T]) get(key interface{}) (T, bool) {
	var zero T
	if len(c.circle) == 0 {
		return zero, false
	}
	hash := c.hashFunction.hash(key)
	if _, ok := c.circle[hash]; !ok {
		keys := c.sortedKeys()
		// tailMap: smallest key >= hash, else wrap to firstKey.
		idx := -1
		for i, k := range keys {
			if k >= hash {
				idx = i
				break
			}
		}
		if idx == -1 {
			hash = keys[0]
		} else {
			hash = keys[idx]
		}
	}
	return c.circle[hash], true
}

func (c *ConsistentHash[T]) sortedKeys() []int {
	keys := make([]int, 0, len(c.circle))
	for k := range c.circle {
		keys = append(keys, k)
	}
	sort.Ints(keys)
	return keys
}
